### PR TITLE
[ROP-7028] Graceful shutdown

### PR DESCRIPTION
**Pull request by Aceeri**
## Description
Waits for the service to handle all incoming requests and after all incoming requests are finished, calling serviceCache.Stop() waits for outgoing requests to finish before shutting down.

Also fixes a small issue in serviceCache where SearchByAction did not lock the index mutex which made concurrent map accesses possible.

## Jira Tickets
https://gudtech.atlassian.net/browse/ROP-7028

### CODE REVIEW

File: scamp/client.go [hunk 1]
```diff
@@ -107,6 +110,26 @@ func (client *Client) Close() {
 		return
 	}
 
+	timeoutPeriod := 30 * time.Second
```
Comment (senseijack) [inline scamp/client.go]: I believe the default timeout when shutting down a docker container is ten seconds. If we're going to add a 30-second timeout we need to make sure we configure Kubernetes to allow for it.

Comment (senseijack) [reply scamp/client.go]: @aceeri Was this PR tested in your dev environment? We should test existing services like SDK to see if they behave correctly. We'll likely have to update SDK and ship-module to remove some of the shutdown logic in those services now that it's been moved to scamp-go. 

I originally thought we'd leave the detection of OS.Signal in the modules themselves and merely update service.Stop() to ensure everything shuts down gracefully when called, I'm not opposed to moving that logic within scamp-go but it somewhat limits things if a future service needs a longer shutdown period for example.

Comment (Aceeri) [reply scamp/client.go]: I can remove the os.Signal from the announcer and we can take the timeout variable as a parameter.

Comment (senseijack) [reply scamp/client.go]: Ok let's do that, and we can have the timeout default to 10 seconds if it's not explicitly set.

**Comment by senseijack** [comment]
I still need to test. Don't merge yet

**Comment by senseijack** [comment]
Closing this PR, we will revisit as part of future scamp enhancements related to autoscaling.

**Pull request closed without merging**